addons/stock_account/tests/test_account_move.py

```diff
@@ -298,3 +298,45 @@ def test_product_valuation_method_change_to_automated_negative_on_hand_qty(self)
             [stock_valuation_line, output_line],
             [expected_valuation_line, expected_output_line]
         )
+
+    def test_stock_account_move_automated_not_standard_with_branch_company(self):
+        """
+        Test that the validation of a stock picking does not fail `_check_company`
+        at the creation of the account move with sub company
+        """
+        main_company = self.env.company
+        branch_a = self.env['res.company'].create({
+            'name': 'Branch A',
+            'country_id': main_company.country_id.id,
+            'parent_id': main_company.id,
+        })
+
+        self.env.user.write({'company_ids': (main_company + branch_a).ids, 'company_id': branch_a.id})
+
+        self.cr.precommit.run()
+
+        categ = self.auto_categ.copy({'property_cost_method': 'average', 'property_valuation': 'real_time'})
+        product = self.product_A
+        product.write({'categ_id': categ.id, 'company_id': branch_a.id, 'standard_price': 300})
+
+        stock_location = self.env['stock.warehouse'].search([
+            ('company_id', '=', self.env.company.id),
+        ], limit=1).lot_stock_id
+
+        in_picking = self.env['stock.picking'].create({
+            'location_id': stock_location.id,
+            'location_dest_id': self.ref('stock.stock_location_customers'),
+            'picking_type_id': stock_location.warehouse_id.in_type_id.id,
+        })
+
+        sm = self.env['stock.move'].create({
+            'name': product.name,
+            'product_id': product.id,
+            'product_uom_qty': 1,
+            'product_uom': product.uom_id.id,
+            'location_id': in_picking.location_id.id,
+            'location_dest_id': in_picking.location_dest_id.id,
+            'picking_id': in_picking.id,
+        })
+        self.assertTrue(in_picking.button_validate())
+        self.assertEqual(self.env.company, sm.account_move_ids.company_id)
```

Comment:
The first value should always be the tested one, and the second value the expected one
```suggestion
        self.assertEqual(sm.account_move_ids.company_id, self.env.company)
```

Reply:
Or even
```suggestion
        self.assertEqual(sm.account_move_ids.company_id, branch_a)
```
To make it really clear